### DISCUSSION TITLE
What’s the memory/RAM overhead of WasmEdge?

### DISCUSSION THREAD

**Discussion by shafqatevo** [WasmEdge/WasmEdge#2228] [General]
Hi, let’s say for a Hello World or an empty wasm module, what is the RAM overhead per each module instance, in terms of KB?

**Comment by alabulei1** [comment]
Hi @shafqatevo 

Sorry for my late reply. We have tested a Wasm app for the Fibonacci number. The peak of memory is about 790 kb. Hope this can help you. Let me know if you have any other questions.

![image](https://user-images.githubusercontent.com/45785633/215384571-9335e7c1-c8b2-429f-a16f-047b7b6c1eca.png)

**Reply by shafqatevo** [reply]
Thanks for this data point, @alabulei1 - really appreciate this!

Best test for RAM overhead measurement is to create hundreds/thousands of no-op/blank module instances and take the median/average incremental RAM overhead for each instantiation. We’re doing the test and will share what we found.

**Reply by alabulei1** [reply]
Got it and thanks for the explanation. Can't wait to see your report!

**Reply by audacioustux** [reply]
https://github.com/audacioustux/playground/blob/main/memtest/wasmedge/src/main.rs hopefully the test code is not too far off from ideal. in summary, 
- it creates 1M modules with simple logic `foo = (n) => n + 1`
- creates 1 instance of each modules (total 1M instance)
- then calls the foo method once for every instance.

and that produces this result: 
<img width="665" alt="Screenshot 2023-01-30 at 1 52 15 PM" src="https://user-images.githubusercontent.com/7238675/215418682-d572911b-93ae-4157-a738-04fd543cbaad.png">

https://raw.githubusercontent.com/audacioustux/playground/main/memtest/massif.out.4384.print

so approximately 3kb per instance

and this is for 1 module only https://github.com/audacioustux/playground/blob/main/memtest/massif.out.8580.print ..
so base memory usage is ~250kb

**Reply by apepkuss** [reply]
Hi @audacioustux, could you please explain what "base memory usage" means? Does it mean the memory consumption after a Vm instance is created? Thanks a lot!

**Reply by audacioustux** [reply]
yes.. basically the memory usage when only 1 instance is created and tested, exactly as described for 1M instances. 
But i'd request to review the code, and give feedback if something is not right, Thanks!
